util.c

```diff
@@ -93,6 +97,192 @@ get_user_cfgfile_path(const char *common_path, const char *filename, const struc
   return 1;
 }
 
+/*
+ * This function will look for users name with valid user token id, in a database Mysql
+ *
+ * Returns one of AUTH_FOUND, AUTH_NOT_FOUND, AUTH_NO_TOKENS, AUTH_ERROR.
+ *
+ * Need database with this table structure : 
+ * 
+ * CREATE TABLE IF NOT EXISTS  `otp`.`yubikey_mappings` ( 
+ *   `otp_id` VARCHAR(12) NOT NULL ,  
+ *   `username` VARCHAR(64) NOT NULL ,     
+ *   PRIMARY KEY  (`otp_id`(12))
+ *  );
+ *
+ */
+int
+check_user_token_mysql (const char *mysql_server,
+    const char *mysql_user,
+    const char *mysql_password,
+    const char *mysql_database,
+		const char *username,
+		const char *otp_id,
+		int verbose,
+    FILE *debug_file)
+{
+
+  //DEFAULT !
+  int retval = AUTH_ERROR;
+  int fd;
+  struct stat st;
+  FILE *opwfile;
+
+  // Mysql
+  MYSQL *con = NULL;
+  MYSQL_STMT *stmt;
+
+  MYSQL_BIND ps_params[2];
+  MYSQL_BIND bind[1];
+
+  long unsigned int username_size = 64;
+  long unsigned int otp_size = 12;
+  unsigned long str_username;
+  unsigned long str_otp;
+  unsigned long length;
+  int int_data;
+  int row_count;
+  int is_null = 0;
+  int error = 0;
+
+  //Check Mysql Librairie
+  if (mysql_library_init(0, NULL, NULL)) {
+    if(verbose)
+	D (debug_file, "could not initialize MySQL client library\n");
+    return retval;
+  }
+
+  con = mysql_init(con);
+  if (!con) {
+    if(verbose)
+	D (debug_file, "out of memorys\n");
+    return retval;
+  }
+
+
+  if (mysql_real_connect(con, mysql_server,mysql_user,mysql_password,mysql_database, 0, NULL, 0) == NULL)
+  {
+    if(verbose)
+	  D (debug_file, "Connection failed ...\n");
+    return retval; 
+  }
+  
+  stmt = mysql_stmt_init(con);
+  if (!stmt)
+    {
+      if(verbose)
+	    D (debug_file, "Connection failed ... 2 \n");
+      return retval; 
+    }
+
+  const char *sql = "SELECT count(username) FROM yubikey_mappings WHERE username = ?;";
+  const char *sql2 = "SELECT count(username) FROM yubikey_mappings WHERE username = ? and otp_id = ?;";
+
+  if(otp_id == NULL)
+  {
+    if (mysql_stmt_prepare(stmt, sql, strlen(sql)))
+    {
+      fprintf(stderr, " mysql_stmt_prepare() failed\n");
+      fprintf(stderr, " %s\n", mysql_stmt_error(stmt));
+      return retval; 
+    }
+  }else{
+    if (mysql_stmt_prepare(stmt, sql2, strlen(sql2)))
+    {
+      fprintf(stderr, " mysql_stmt_prepare() failed\n");
+      fprintf(stderr, " %s\n", mysql_stmt_error(stmt));
+      return retval; 
+    }
+  }
+
+
+  str_username= strlen(username);  
+  memset(ps_params, 0, sizeof(ps_params));
+  ps_params[0].buffer_type = MYSQL_TYPE_STRING;
+  ps_params[0].buffer = (char *)username;
```

Comment:
What's the guarantees here? would it be better to copy username to a local buffer and use that? Especially since we claim that the buffer is 64 bytes here.

Reply:
I have change this party by :

```
  str_username = strlen(username);  
  memset(ps_params, 0, sizeof(ps_params));
  ps_params[0].buffer_type = MYSQL_TYPE_STRING;
  ps_params[0].buffer = (char *)username;
  ps_params[0].buffer_length = str_username;
  ps_params[0].length = &str_username;
  ps_params[0].is_null = 0;
```

The mysql documentation is not really clear about the usage of length et buffer_length.
If you really want a local buffer, I can do it. But I prefer like I have wrote.

Reply:
that const has to be thrown away on username might be a hint. I'm fine with this if it seems to be how others use this?